File: js/packages/app/src/__tests__/features/global-functions/wrappers/global-function-example/src/polywrap/external/module/InternalWrapInstance.ts
import { IExternalWrapInstance, IInternalWrapInstance } from "@nerfzael/reim-wrap-as";
import { invoke as invokeGlobalFunction } from "../../internal/global-functions/invokeGlobalFunction";
import { invokeClassMethod } from "../../internal/classes/invokeClassMethod";
import { InternalResource } from "../../dt/InternalResource";

export class InternalWrapInstance extends IInternalWrapInstance {
  constructor() {
    super();
  }

  invokeResource(resource: u32, buffer: ArrayBuffer, externalWrapInstance: IExternalWrapInstance): ArrayBuffer {
    switch (resource) {
      case InternalResource.InvokeGlobalFunction:
        return invokeGlobalFunction(buffer, externalWrapInstance);
      case InternalResource.InvokeClassMethod: 
        return invokeClassMethod(buffer, externalWrapInstance);
      default:
        throw new Error("Unknown resource: " + resource.toString());
    }
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/external/module/InternalWrapInstance.rs
use reim_wrap::{ ExternalWrapModule, InternalWrapModule };
use internal::global_functions::{ invoke as invoke_global_function };
use internal::classes::{ invoke_class_method };
use dt::{ InternalResource };

struct InternalModule {}

impl InternalModule {
    pub fn new() -> Self {
        InternalModule {}
    }
}

impl InternalWrapModule for InternalModule {
  async fn invoke_resource(&self, resource: u32, buffer: &[u8], external_module: dyn ExternalWrapModule) -> Vec<u8> {
    match resource {
      InternalResource::InvokeGlobalFunction => invoke_global_function(buffer, external_module),
      InternalResource::InvokeClassMethod => invoke_class_method(buffer, external_module),
      _ => panic!("Unknown resource: ".to_string() + &resource.to_string()),
    }
  }
}
